File: pkg/engine/patterns.go
package engine

import (
	"math"
	"strings"
)

// pattern is a named traffic shaping function.
type pattern int

const (
	patternSteady   pattern = iota
	patternBursty
	patternDiurnal
	patternIncident
)

// matchPattern maps a free-form traffic_pattern string to a known pattern.
func matchPattern(s string) pattern {
	s = strings.ToLower(s)
	switch {
	case strings.Contains(s, "bursty") || strings.Contains(s, "burst"):
		return patternBursty
	case strings.Contains(s, "diurnal"):
		return patternDiurnal
	case strings.Contains(s, "incident"):
		return patternIncident
	default:
		return patternSteady
	}
}

// multiplier returns the traffic scaling factor (≥0) for the given tick.
// rng is used for jitter.
func multiplier(p pattern, tickIndex int, rng interface{ Float64() float64 }) float64 {
	switch p {
	case patternBursty:
		// Spike every ~30 ticks; 5× peak, 0.2× trough.
		phase := float64(tickIndex%30) / 30.0
		base := 0.2 + 0.8*(1+math.Sin(2*math.Pi*phase))/2
		jitter := 1 + (rng.Float64()-0.5)*0.3
		return base * jitter

	case patternDiurnal:
		// One full day per 86400 ticks; peak at noon (tick 43200), trough at 3am.
		hour := float64(tickIndex%86400) / 3600.0
		// Shifted cosine: 0 at 3am (hour 3), peak at 3pm (hour 15).
		v := (1 - math.Cos(2*math.Pi*(hour-3)/24)) / 2
		v = 0.05 + 0.95*v
		jitter := 1 + (rng.Float64()-0.5)*0.1
		return v * jitter

	case patternIncident:
		// normal(0–99) → spike(100–149) → degraded(150–299) → recovery(300–399)
		phase := tickIndex % 400
		var v float64
		switch {
		case phase < 100:
			v = 1.0
		case phase < 150:
			v = 1.0 + float64(phase-100)/50.0*4.0 // ramp up to 5×
		case phase < 300:
			v = 5.0
		default:
			v = 5.0 - float64(phase-300)/100.0*4.0 // ramp back down
		}
		jitter := 1 + (rng.Float64()-0.5)*0.2
		return v * jitter

	default: // steady
		jitter := 1 + (rng.Float64()-0.5)*0.1
		return jitter
	}
}

File: src/engine/traffic/patterns.ts
export type TrafficPattern = 'steady' | 'bursty' | 'diurnal' | 'incident'

/**
 * Get traffic multiplier based on pattern and tick context
 */
export function getTrafficMultiplier(
  pattern: TrafficPattern,
  tickIndex: number,
  rng: () => number
): number {
  switch (pattern) {
    case 'steady':
      return 1 + (rng() - 0.5) * 0.2 // ±10% noise

    case 'bursty': {
      // Every ~20 ticks, a burst lasting 3-5 ticks
      const burstPeriod = 20
      const inBurst = (tickIndex % burstPeriod) < 5
      return inBurst ? 3 + rng() * 4 : 0.3 + rng() * 0.4
    }

    case 'diurnal': {
      // Simulate time-of-day: peak at hour 9-17 (assuming 1 tick = 1 second)
      const secondsInDay = 86400
      const secondOfDay = tickIndex % secondsInDay
      const hourOfDay = secondOfDay / 3600
      // Gaussian-ish peak at midday
      const peakHour = 13
      const spread = 4
      const base = Math.exp(-0.5 * Math.pow((hourOfDay - peakHour) / spread, 2))
      const noise = 1 + (rng() - 0.5) * 0.1
      return 0.1 + base * 0.9 * noise
    }

    case 'incident': {
      // Phases: normal (0-30%), spike (30-50%), errors (50-80%), recovery (80-100%)
      const phase = (tickIndex % 100) / 100
      if (phase < 0.3) return 1 + (rng() - 0.5) * 0.2
      if (phase < 0.5) return 3 + rng() * 5
      if (phase < 0.8) return 2 + rng() * 2
      return 0.5 + (phase - 0.8) / 0.2 * 0.5
    }

    default:
      return 1
  }
}
